### PR TITLE
Linkification

### PR DESCRIPTION
**Pull request by Daiz**
This pull request is all about linkification. SHA linkification is partially done already.
- [x] @mention
- [x] #Num
- [x] User/#Num
- [x] User/Project#Num
- [x] SHA
- [x] User@SHA
- [x] User/Project@SHA

### CODE REVIEW

File: src/linkify.ls [hunk 1]
```diff
@@ -0,0 +1,17 @@
+Linkify =
+  sha1: (text, context) ->
+    regex = /([A-Za-z0-9-]+)?\/?([A-Za-z0-9_-]+)?@?([a-f0-9]{40})/
+    match = text.match regex
+    user = match.1
+    repo = match.2
+    hash = match.3
+    short = hash.substr 0 8
+    ctx = context.split '/'
```
Comment (vendethiel) [inline src/linkify.ls]: `[, ctx] =` ?

File: src/linkify.ls [hunk 2]
```diff
@@ -0,0 +1,17 @@
+Linkify =
+  sha1: (text, context) ->
+    regex = /([A-Za-z0-9-]+)?\/?([A-Za-z0-9_-]+)?@?([a-f0-9]{40})/
+    match = text.match regex
```
Comment (vendethiel) [inline src/linkify.ls]: `[user, repo, hash] =` ?

**Comment by Daiz** [comment]
```
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

What is this I don't even

**Comment by vendethiel** [comment]
Github sometimes has little problems

**Comment by Daiz** [comment]
Both issue linkification and @mention linkification have some issues at the moment. SHA linkification seems to mostly work. Think I'll get back to this tomorrow.

**Comment by Daiz** [comment]
E-mails should also be automatically linkified.

File: Slakefile [hunk 3]
```diff
@@ -17,14 +22,14 @@ task \bundle 'Bundle dependencies.' !->
   b.require \marked
   b.require \highlight.js
   (err, deps) <-! b.bundle {}
-  deps .to 'vendor.js'
```
Comment (vendethiel) [inline Slakefile:20]: Parens around args

Comment (Daiz) [inline Slakefile:20]: I'll get rid of those eventually.

Comment (vendethiel) [inline Slakefile:20]: I'll leave more styling comments when I'm no more on my phone

File: src/head.ls [hunk 4]
```diff
@@ -0,0 +1,10 @@
+require! {
+  \marked
+  hljs: \highlight.js
+}
+
+# shortcuts
+d = document
+context = do ->
```
Comment (vendethiel) [inline src/head.ls:8]: `let`

File: src/linkify.ls [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  linkify =
+    all: (text, context) ->
+      text = @sha text, context
+      text = @issue text, context
+      # text = @mention text
+  
+    sha: (text, context, ret = '') ->
+      start = end = 0
+      regex = /([A-Za-z0-9-]+)?\/?([A-Za-z0-9_-]+)?@?([a-f0-9]{40})/g
+      while res = regex.exec text
```
Comment (vendethiel) [inline src/linkify.ls]: you have `that`

File: src/linkify.ls [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        [matched, user, repo, hash] = res
+        start = end
+        end = res.index - start + matched.length
+        current = text.substr start, end
+        rest = text.substr end
+        short = hash.substr 0 8
+        ctx-user = context.split '/' .0
+        ret += current.replace matched, if user == ctx-user and not repo
```
Comment (vendethiel) [inline src/linkify.ls]: switch?

**Comment by Daiz** [comment]
This thing sure has a lot of edge cases. Thank god for unit testing.

**Comment by Daiz** [comment]
SHA linkification finally passes all the tests! :tada:

Guess I should start dealing with issue linkification next.

**Comment by Daiz** [comment]
Oh boy, more edge cases! Because everybody _loves_ edge cases.

**Comment by Daiz** [comment]
Man, this has almost as much commits as `master` does at the moment.

**Comment by Daiz** [comment]
Yeah, I think this branch has enough content for merging by now.
